File: src/workflow/graph_factory.py
"""
工作流圖工廠 (Workflow Graph Factory)

統一創建和管理不同領域的工作流圖。
支持動態加載適配器和配置。
"""

from typing import Optional, Dict, Any
from enum import Enum

from src.workflow.adapters import BaseDomainAdapter, AdapterRegistry


class WorkflowLevel(Enum):
    """工作流層級 (CCW 風格)"""
    L1_RAPID = 1          # 快速模式 - 無狀態
    L2_LIGHTWEIGHT = 2    # 輕量模式 - 內存計劃
    L3_STANDARD = 3       # 標準模式 - 完整會話
    L4_BRAINSTORM = 4     # 頭腦風暴 - 多角色並行
    L5_COORDINATOR = 5    # 智能編排 - 自動規劃


class WorkflowFactory:
    """
    工作流圖工廠
    
    用法:
        # 創建小說工作流
        graph = WorkflowFactory.create("novel", config={"pass_score": 85})
        
        # 創建代碼工作流
        graph = WorkflowFactory.create("code", level=WorkflowLevel.L3_STANDARD)
        
        # 列出可用領域
        domains = WorkflowFactory.list_domains()
    """
    
    @staticmethod
    def create(
        domain: str,
        level: WorkflowLevel = WorkflowLevel.L3_STANDARD,
        config: Optional[Dict[str, Any]] = None
    ):
        """
        創建工作流圖
        
        Args:
            domain: 領域類型 ("novel" | "code" | "knowledge" | ...)
            level: 工作流層級 (L1-L5)
            config: 自定義配置
            
        Returns:
            編譯後的 LangGraph 應用
            
        Raises:
            ValueError: 如果領域未註冊
        """
        adapter = AdapterRegistry.create_adapter(domain, config)
        
        if adapter is None:
            available = AdapterRegistry.list_domains()
            raise ValueError(
                f"Unknown domain: '{domain}'. "
                f"Available domains: {available}"
            )
        
        # 合併配置
        merged_config = adapter.merge_config(config)
        merged_config["workflow_level"] = level.value
        
        # 創建圖
        graph = adapter.create_graph()
        
        return graph
    
    @staticmethod
    def create_adapter(
        domain: str,
        config: Optional[Dict[str, Any]] = None
    ) -> Optional[BaseDomainAdapter]:
        """
        創建適配器實例 (不編譯圖)
        
        用於需要訪問適配器方法但不需要完整工作流的場景
        """
        return AdapterRegistry.create_adapter(domain, config)
    
    @staticmethod
    def list_domains() -> list[str]:
        """列出所有已註冊的領域"""
        return AdapterRegistry.list_domains()
    
    @staticmethod
    def register_adapter(domain: str, adapter_class: type):
        """
        手動註冊適配器
        
        用於動態加載自定義領域
        """
        AdapterRegistry._adapters[domain] = adapter_class
    
    @staticmethod
    def get_level_description(level: WorkflowLevel) -> Dict[str, str]:
        """獲取層級描述"""
        descriptions = {
            WorkflowLevel.L1_RAPID: {
                "name": "Rapid",
                "description": "快速模式 - 無狀態、無工件、直接輸出",
                "novel_use": "錯字修正、格式調整、快速潤色",
                "code_use": "typo 修復、格式化、簡單重命名",
            },
            WorkflowLevel.L2_LIGHTWEIGHT: {
                "name": "Lightweight",
                "description": "輕量模式 - 內存計劃、輕量持久化",
                "novel_use": "單章節寫作、情節調整",
                "code_use": "單文件修改、簡單功能添加",
            },
            WorkflowLevel.L3_STANDARD: {
                "name": "Standard",
                "description": "標準模式 - 完整會話、驗證步驟",
                "novel_use": "多章節開發、角色塑造",
                "code_use": "功能開發 + 測試、代碼審查",
            },
            WorkflowLevel.L4_BRAINSTORM: {
                "name": "Brainstorm",
                "description": "頭腦風暴 - 多角色並行分析",
                "novel_use": "世界觀設計、劇情構思",
                "code_use": "架構設計、技術選型",
            },
            WorkflowLevel.L5_COORDINATOR: {
                "name": "Coordinator",
                "description": "智能編排 - 自動規劃命令鏈",
                "novel_use": "完整小說創作、長期項目",
                "code_use": "完整項目開發、多模塊協調",
            },
        }
        return descriptions.get(level, {})


# ============================================================
# 便捷函數
# ============================================================

def create_workflow(
    domain: str,
    user_request: str,
    level: int = 3,
    config: Optional[Dict] = None
):
    """
    便捷函數: 創建並初始化工作流
    
    Args:
        domain: 領域類型
        user_request: 用戶請求
        level: 工作流層級 (1-5)
        config: 自定義配置
        
    Returns:
        (graph, initial_state) 元組
    """
    workflow_level = WorkflowLevel(level)
    
    adapter = WorkflowFactory.create_adapter(domain, config)
    if adapter is None:
        raise ValueError(f"Unknown domain: {domain}")
    
    graph = adapter.create_graph()
    initial_state = adapter.create_initial_state(user_request)
    
    return graph, initial_state

File: src-ts/workflow/graph-factory.ts
/**
 * Graph Factory - Unified creation of configured workflow graphs
 *
 * Migrated from src/workflow/graph_factory.py
 */

import { workflowLevelFromLabel, type WorkflowLevelValue } from './types';
import { AdapterRegistry, type BaseDomainAdapter } from './adapters/base-adapter';
import { determineResumeStrategy, type ResumeDecision } from './session/resume-strategy';

// ============================================================
// Helper functions
// ============================================================

function buildResumeDecision(
  config: Record<string, unknown> | null,
  tool: string,
  resumeIds?: string[],
  customId?: string | null,
): ResumeDecision | null {
  if (!config || !resumeIds) return null;

  return determineResumeStrategy(
    tool,
    resumeIds,
    customId ?? undefined,
  );
}

function mergeResumeMetadata(
  metadata: Record<string, unknown>,
  resumeDecision: ResumeDecision | null,
): Record<string, unknown> {
  const merged: Record<string, unknown> = { ...metadata };
  if (resumeDecision) {
    if (!merged['resume_decision']) {
      merged['resume_decision'] = resumeDecision;
    }
  }
  return merged;
}

// ============================================================
// WorkflowFactory
// ============================================================

export class WorkflowFactory {
  static create(
    domain: string,
    level: WorkflowLevelValue | string | number = 3,
    config?: Record<string, unknown> | null,
  ): unknown {
    let workflowLevel: WorkflowLevelValue;
    if (typeof level === 'string' || typeof level === 'number') {
      workflowLevel = workflowLevelFromLabel(level);
    } else {
      workflowLevel = level;
    }

    const adapter = AdapterRegistry.createAdapter(domain, config ?? null);
    if (!adapter) {
      const available = AdapterRegistry.listDomains();
      throw new Error(
        `Unknown domain: '${domain}'. Available domains: ${available.join(', ')}`,
      );
    }

    const mergedConfig = adapter.mergeConfig(config ?? null) as Record<string, unknown>;
    mergedConfig['workflow_level'] = typeof workflowLevel === 'number' ? workflowLevel : 3;

    const configuredAdapter = AdapterRegistry.createAdapter(domain, mergedConfig);
    return (configuredAdapter ?? adapter).createGraph();
  }

  static createAdapter(
    domain: string,
    config?: Record<string, unknown> | null,
  ): BaseDomainAdapter | null {
    return AdapterRegistry.createAdapter(domain, config ?? null);
  }

  static listDomains(): string[] {
    return AdapterRegistry.listDomains();
  }

  static registerAdapter(
    domain: string,
    adapterClass: new (...args: unknown[]) => BaseDomainAdapter,
    capabilities?: string[] | null,
  ): void {
    AdapterRegistry.registerAdapter(domain, adapterClass, capabilities ?? undefined);
  }

  static listDomainsByCapability(capability: string): string[] {
    return AdapterRegistry.listDomainsByCapability(capability);
  }

  static getAdapterCapabilities(domain: string): string[] {
    return AdapterRegistry.getCapabilities(domain);
  }

  static getLevelDescription(level: number): Record<string, string> {
    const descriptions: Record<number, Record<string, string>> = {
      1: {
        name: 'Rapid',
        description: '快速模式 - 无状态、无工件、直接输出',
        novel_use: '错字修正、格式调整、快速润色',
        code_use: 'typo 修复、格式化、简单重命名',
      },
      2: {
        name: 'Lightweight',
        description: '轻量模式 - 内存计划、轻量持久化',
        novel_use: '单章节写作、情节调整',
        code_use: '单文件修改、简单功能添加',
      },
      3: {
        name: 'Standard',
        description: '标准模式 - 完整会话、验证步骤',
        novel_use: '多章节开发、角色塑造',
        code_use: '功能开发 + 测试、代码审查',
      },
      4: {
        name: 'Brainstorm',
        description: '头脑风暴 - 多角色并行分析',
        novel_use: '世界观设计、剧情构思',
        code_use: '架构设计、技术选型',
      },
      5: {
        name: 'Coordinator',
        description: '智能编排 - 自动规划命令链',
        novel_use: '完整小说创作、长期项目',
        code_use: '完整项目开发、多模块协调',
      },
    };
    return descriptions[level] ?? {};
  }
}

// ============================================================
// Convenience function
// ============================================================

export function createWorkflow(
  domain: string,
  userRequest: string,
  level: number = 3,
  config?: Record<string, unknown> | null,
): { graph: unknown; initialState: Record<string, unknown> } {
  const cfg: Record<string, unknown> = { ...(config ?? {}) };
  if (!('workflow_level' in cfg)) {
    cfg['workflow_level'] = level;
  }

  const adapter = WorkflowFactory.createAdapter(domain, cfg);
  if (!adapter) {
    throw new Error(`Unknown domain: ${domain}`);
  }

  const graph = adapter.createGraph();

  const resumeDecision = buildResumeDecision(
    cfg,
    (cfg['tool'] as string) ?? 'gemini',
    cfg['resume_ids'] as string[] | undefined,
    (cfg['custom_id'] as string) ?? null,
  );

  const baseMetadata =
    typeof cfg['metadata'] === 'object' && cfg['metadata'] !== null
      ? (cfg['metadata'] as Record<string, unknown>)
      : {};
  const metadata = mergeResumeMetadata(baseMetadata, resumeDecision);

  const initialState = adapter.createInitialState(userRequest, {
    metadata,
    resume_decision: resumeDecision,
  });

  return { graph, initialState: initialState as unknown as Record<string, unknown> };
}
